### PR TITLE
Update dstore/Cache::put

### LINKED ISSUE #117: Cache.js deletes element while waiting: necessary?
Hi,

Cache.js has the following lines:

```
 put: function (object, directives) {
        // first remove from the cache, so it is empty until we get a response from the master store
        var cachingStore = this.cachingStore;
        cachingStore.remove((directives && directives.id) || this.getIdentity(object));
        return when(this.inherited(arguments), function (result) {
```

That `cachingStore.remove` is giving me a lot of grief, because I am listening to `update` events:

```
 store.on( 'add,update,delete', function( event ){
            var first = store.memCache.data[ 0 ];
```

And the problem is that when the listener runs, the cache is effectively missing the element (and I need it to be there).

Why are we deleting that element? This is what used to happen with Dojo's store... but is it strictly necessary? Isn't it a borderline bug, to have an incomplete cache for cached stores? Is there any chance we could make that deletion optional? (it really throws a spanner in the works for me)

### PR DESCRIPTION
**Pull request by kitsonk**
Previous behaviour was to remove the item while waiting for return from the master store.  Now Cache will presume the update will complete and only remove the item if the master store rejects the `put`.

Fixes #117.

### CODE REVIEW

File: src/Cache.js [hunk 1]
```diff
@@ -135,16 +135,20 @@ define([
...
+			// first update the cache, we are going to assume the update is valid while we wait to
+			// hear from the master store
 			var cachingStore = this.cachingStore;
-			cachingStore.remove((directives && directives.id) || this.getIdentity(object));
-			return when(this.inherited(arguments), function (result) {
+			cachingStore.put(object, directives);
+			return when(this.inherited(arguments)).then(function (result) {
 				// now put result in cache
 				var cachedPutResult =
 					cachingStore.put(object && typeof result === 'object' ? result : object, directives);
```
Comment (kfranqueiro) [inline src/Cache.js]: This can be reduced to an `if`, since if result is not an object, we've already done this prior to calling inherited.

**Comment by kriszyp** [comment]
This looks good to me.

**Comment by dylans** [comment]
I think this one needs a rebase. We have enough things to do a 1.2 release soon so we should rebase and land this if time permits @kitsonk.

**Comment by kitsonk** [comment]
Yeah, the rebase totally screwed up... :cry: This is simply too stale.

**Comment by kitsonk** [comment]
A cherry-pick to a new branch worked.

**Pull request merged**